### ISSUE TITLE
TS compiler exits unexpectly while trying to report an error

### ISSUE THREAD

**Issue by alexsaveliev** [Microsoft/TypeScript#15128]
**TypeScript Version:**  Version 2.3.0-dev.20170411

Seen with 2.2.2 and 2.2.1 too, haven't checked older versions

- clone https://github.com/Microsoft/TypeScript somewhere
- remove all `tsconfig.json` files from cloned copy and put the following one in root folder:
```
{
  "compilerOptions": {
    "outDir": "lib"
  },
  "include": [
    "**/*.ts"
  ],
  "exclude": [
    "node_modules/**/*"
  ]
}
```
(yeah, I know that tje project shouldn't compile this way but probably there are real cases where error might occur)
- run `tsc`

**Actual behavior:**

TSC tries to report compilation error but fails with

```
C:\Users\alexsaveliev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:53644
                throw e;
                ^

TypeError: Cannot read property 'name' of undefined
    at mergeSymbol (C:\Users\alexsaveliev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:21057:46)
    at C:\Users\alexsaveliev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:21081:21
    at Map.forEach (native)
    at mergeSymbolTable (C:\Users\alexsaveliev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:21071:20)
    at initializeTypeChecker (C:\Users\alexsaveliev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:38187:21)
    at Object.createTypeChecker (C:\Users\alexsaveliev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:20938:9)
    at getDiagnosticsProducingTypeChecker (C:\Users\alexsaveliev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:53551:93)
    at emitWorker (C:\Users\alexsaveliev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:53584:32)
    at C:\Users\alexsaveliev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:53560:66
    at runWithCancellationToken (C:\Users\alexsaveliev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:53637:24)
```
matching snippet of `tsc.js` is:

```
                error(source.valueDeclaration.name, ts.Diagnostics.Cannot_augment_module_0_with_value_exports_because_it_resolves_to_a_non_module_entity, symbolToString(target));
```

**Comment by sandersn** [comment]
This is a pretty easy fix; find a real declaration instead of `valueDeclaration`, which is sometimes missing. What's harder is writing a test. I'll debug to the crash to see what source causes the crash and try to get a small repro from there.

**Comment by sandersn** [comment]
It happens because we're compiling a bunch of our tests, which all define `namespace A { ... }` that merge together. The crash happens trying to report on the error for merging `type A<T extends Unknown> = {}`. 

It makes sense that `type A` doesn't have a valueDeclaration since it has no values.

**Comment by sandersn** [comment]
Here's the repro:

```ts
// @Filename: initial.ts
interface A { }
namespace A { }
// @Filename: final.ts
type A = { }
```

First merge an interface and a namespace in one file. Then try to merge a type alias in a separate file.

**Comment by sandersn** [comment]
Probably the reason we haven't seen this before is that people rarely intentionally merge type aliases with anything else. Fix is up at #15258
